subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultConvention.java

```diff
@@ -91,44 +105,68 @@ private Instantiator getInstantiator() {
         return values.get(0);
     }
 
+    @Override
     public void add(String name, Object extension) {
         if (extension instanceof Class) {
             create(name, (Class<?>) extension);
         } else {
-            extensionsStorage.add(name, extension);
+            add(name, (Class) extension.getClass(), extension);
         }
     }
 
+    @Override
+    public <P, I extends P> void add(String name, Class<P> publicType, I extension) {
```

Comment:
I can see that `I` is meant to be mean `InstanceType` here but it might be better to use `T` for the concrete type of the extension in all overloads, what do you think?